src/cpp/src/visual_language/pipeline.cpp

```diff
@@ -739,6 +1016,45 @@ class ov::genai::VLMPipeline::VLMPipelineImpl {
         pipe.m_resampler.infer();
         return pipe.m_resampler.get_output_tensor();  // [N, query_num, new_hidden_size]
     }
+
+    ov::Tensor get_encoded_input_ids(const std::string& prompt, const std::string& chat_template_fallback = "") {
+        ov::Tensor encoded_input_ids;
+        if (m_is_chat_conversation) {
+            // KV cache in model already contains prompts and answers from previous iterations.
+            // So only new prompt wrapped into chat template to be sent into model. Tokenizer always returns
+            // token_ids = {<bos token>, ...<valuable tokens>}. So if tokenizer applies only to the new prompt,
+            // <bos token> will be inserted on every iteration.
+            // So actual pipeline calculates input_ids for whole chat history + for whole chat history without the new prompt
+            // and takes only the difference between them.
+            // The chat history cannot be saved as already encoded tokens because generate call doesn't return <eos> token, but
+            // KV cache contains it. So we have to add it manually or get it by tokenization all chat history.
+            m_history.push_back({{"role", "user"}, {"content", prompt}});
+            constexpr bool add_generation_prompt = true;
+            std::string new_templated_chat_history;
+            try {
+                new_templated_chat_history = m_tokenizer.apply_chat_template(m_history, add_generation_prompt);
+            } catch (const std::exception& error) {
+                // Use fallback chat template if it was not found in tokenizer_config.json
+                new_templated_chat_history = m_tokenizer.apply_chat_template(m_history, add_generation_prompt, chat_template_fallback);
+            }
+            ov::Tensor new_chat_tokens = m_tokenizer.encode(new_templated_chat_history).input_ids;
+            auto history_len = m_language.get_tensor("attention_mask").get_shape().at(1);
+            if (history_len == 0) {
+                encoded_input_ids = new_chat_tokens;
+            } else {
+                TokenizedInputs prev_chat_tokens = m_tokenizer.encode(
+                    m_templated_chat_history
+                );
+                encoded_input_ids = utils::subtract_chat_tokenized_inputs(
+                    {new_chat_tokens}, prev_chat_tokens
+                ).input_ids;
+            }
+            m_templated_chat_history = std::move(new_templated_chat_history);
+        } else {
+            encoded_input_ids = m_tokenizer.encode(prompt).input_ids;
```

Comment:
@Wovchena @ilya-lavrenov Does it make sense to use `chat_template_fallback` (if provided) instead of `prompt` in **not** chat scenario as well?

Reply:
No. Enable chat teplate only for `start_chat()` to stay aligned with LLMPipeline and provide more ways to process the prompt.